### ISSUE TITLE
Custom forms directory

### ISSUE THREAD

**Issue by xylo04** [la5nta/pat#374]
There have been a couple of questions asking about support for custom forms similar to [RMS Express](https://winlink.org/content/forms_not_standard_library_download_process_instructions). This should just consist of adding another directory to scan for forms. It was reported that adding custom forms to the standard directory doesn't work; I'm surprised, that bears investigation. However, a separate directory is still needed so custom forms don't get deleted when upgrading standard forms.

**Comment by martinhpedersen** [comment]
Agreed 👍

While we're at it, maybe we could make it a list so we can support arbitrary number of forms bundles?

**Comment by martinhpedersen** [comment]
>  It was reported that adding custom forms to the standard directory doesn't work; I'm surprised, that bears investigation. 

I did some digging. The user that reported this did not clarify if they attempted to add a HTML Form Template or a pure .txt template. I'm guessing the latter. As it turns out, there is a big difference.. and our forms package only supports HTML Forms. We ignore pure txt templates.

It looks to me like HTML Forms is a more formalized and feature rich version of an older(?) Message Templates feature.

The differences between a pure txt template (often referred to as "custom templates"?) and HTML Forms needs to be investigated.

* [Message Templates](https://winlink.org/content/winlink_express_message_templates)
* [HTML Forms](https://winlink.org/content/winlink_express_html_forms)

You can tell the difference by looking for the `Form:` line at the beginning of the .txt.

**Comment by Famman49** [comment]
The form I had problems with was an HTML form.

**Comment by richcannings** [comment]
@xylo04 and @Famman49: Send me a link to the forms you are having challenges with, and even better, the steps to reproduce the issue.

I've been tinkering with the forms system and fixed an issue with NorCal Winlink Net forms (#384).

**Comment by sarexpert** [comment]
*pandoc* will convert almost any text or wisiwg document to html.  It is a command line program.

https://pandoc.org/

**Comment by philippbike** [comment]
I can confirm that I am able to upload custom forms into the standard template directory that the WinLink forms are located, and they are usable from within pat. But if the WinLink forms are re-pulled/updated, pat ends up wiping anything custom from that directory, so a specific custom forms folder separate of the WinLink forms would definitely save me alot of hassle :P

**Comment by MarkMartinec** [comment]
If a html custom form contains a non-ASCII character, pat ignores and it does not show it in the list of available templates. This kills usability of templates for international audience, especially as template data (like attachments) are transferred 8-bit clean (unlike the main message due to limitation of the B2F protocol).

The Winlink Express works just fine with UTF-8 character set in html templates. I consider it a bug in pat.

**Comment by martinhpedersen** [comment]
Can you please provide an example? UTF-8 is the intended supported/expected character set.

**Comment by martinhpedersen** [comment]
Oh, and please file a new issue. Your UTF-8 issue is not related to this (Custom forms directory).

**Comment by MarkMartinec** [comment]
> If a html custom form contains a non-ASCII character, pat ignores it

After additional testing - I retract my claim, sorry.

The full story: I have installed a set of files for a custom template in the .../pat/Standard_Forms/General Forms/ directory (with the same ownership and permissions as other forms). Even after restarting the pat the new form would not show in the list of available forms, hard-reloading the GUI page in the web browser did not help either.

After reading Dominic OE7DRT on Feb 24, 2025 [https://groups.google.com/g/pat-users/c/SvT0pXEZtZw/m/my-groups](https://groups.google.com/g/pat-users/c/SvT0pXEZtZw/m/my-groups), who found out that a degree sign ° in some weather forms showing °C was a culprit for his inability to load forms, I tried to replace all non-ASCII UTF-8 characters in my forms (i.e. our Slovenian letters č, š, ž) with an ASCII replacement - and after some mucking around the form finally did load - so this is when I posted my comment above.

But on today's re-testing I realized that my original custom forms with UTF-8 characters does indeed show in the list and does work as intended. Turns out there was some caching involved in the web browser (pat GUI), and restarting pat and hard-refresh of the web page did not help - but ditching the tab and reopening the GUI did indeed update the list of templates. Possibly something is cached in cookies or the like. This was a chromium web browser.

So to recap: to re-load the set of pat templates in GUI, re-opening of the GUI tab is necessary (not just a page hard-refresh). Restarting pat is not enough.
